File: tests/oracle/block_identify_internal_1204.cc
/*
 * BLOCKSTRUCT-IDENTIFY-BOUNDARY-0001 locked Ghidra 12.0.4 oracle.
 *
 * Directly exercises BlockGraph::identifyInternal + selfIdentify + dedup
 * (block.cc:940-963 / 895-931 / 525-539) through the public factory
 * BlockGraph::newBlockList (block.cc:1758-1774).  Observed state: composite
 * children order, component parent ownership, raw block flags (components
 * are NEVER f_dead), inherited boundary edge slots/labels/reverse indices
 * on the composite, retargeted peer halves, parallel-edge dedup label
 * merge, FlowBlock::f_switch_out / f_interior_goto* flag propagation, and internal
 * (component-to-component) edge retention.
 *
 * Top-level list positions are normalized to sorted membership: Ghidra
 * appends the composite at the end of the graph list (block.cc:1695),
 * Rugra installs it at the first component's slot — a registered model
 * divergence outside this fixture's covered projection.  Every other
 * observation is printed raw.
 */
#include <bits/stdc++.h>

#define private public
#define class struct
#include "block.hh"
#undef class
#undef private

namespace {

using namespace ghidra;

class Names {
  std::map<const FlowBlock *, std::string> blocks;
public:
  void add(const FlowBlock *ptr,const std::string &name) { blocks[ptr] = name; }

  std::string block(const FlowBlock *ptr) const
  {
    if (ptr == (const FlowBlock *)0) return "null";
    std::map<const FlowBlock *,std::string>::const_iterator iter = blocks.find(ptr);
    return iter == blocks.end() ? "?" : iter->second;
  }

  std::string parent(const FlowBlock *ptr) const { return block(ptr->getParent()); }

  std::string sortedMembers(const std::vector<FlowBlock *> &list) const
  {
    std::vector<std::string> names;
    for(int4 i=0;i<list.size();++i)
      names.push_back(block(list[i]));
    std::sort(names.begin(),names.end());
    std::ostringstream result;
    result << '[';
    for(int4 i=0;i<names.size();++i) {
      if (i != 0) result << ',';
      result << names[i];
    }
    result << ']';
    return result.str();
  }
};

void setIndex(FlowBlock *bl,int4 index)
{
  bl->index = index;
}

void addLabeledEdge(BlockGraph &graph,FlowBlock *source,FlowBlock *target,uint4 label)
{
  int4 outslot = source->outofthis.size();
  int4 inslot = target->intothis.size();
  graph.addEdge(source,target);
  source->outofthis[outslot].label = label;
  target->intothis[inslot].label = label;
}

std::string edgeList(const FlowBlock *block,bool outgoing,const Names &names)
{
  const std::vector<BlockEdge> &edges = outgoing ? block->outofthis : block->intothis;
  std::ostringstream result;
  result << '[';
  for(int4 slot=0;slot<edges.size();++slot) {
    if (slot != 0) result << ',';
    result << names.block(edges[slot].point) << ':'
           << edges[slot].reverse_index << ':' << edges[slot].label;
  }
  result << ']';
  return result.str();
}

// Reciprocal-consistency check mirroring FlowBlock::checkEdges
// (block.cc:545-570) without the BLOCKCONSISTENT_DEBUG guard.
bool edgesConsistent(const FlowBlock *bl)
{
  for(int4 i=0;i<bl->intothis.size();++i) {
    const BlockEdge &edge(bl->intothis[i]);
    int4 rev = edge.reverse_index;
    const FlowBlock *peer = edge.point;
    if (rev < 0 || peer->outofthis.size() <= rev) return false;
    if (peer->outofthis[rev].point != bl) return false;
    if (peer->outofthis[rev].reverse_index != i) return false;
  }
  for(int4 i=0;i<bl->outofthis.size();++i) {
    const BlockEdge &edge(bl->outofthis[i]);
    int4 rev = edge.reverse_index;
    const FlowBlock *peer = edge.point;
    if (rev < 0 || peer->intothis.size() <= rev) return false;
    if (peer->intothis[rev].point != bl) return false;
    if (peer->intothis[rev].reverse_index != i) return false;
  }
  return true;
}

// Case 1: plain cat identify. P -> A -> B -> C -> Q, plus an external
// in-edge P -> A with label, consumed set {A,B}.
void caseCatBasic()
{
  BlockGraph g;
  FlowBlock *P = g.newBlock();
  FlowBlock *A = g.newBlock();
  FlowBlock *B = g.newBlock();
  FlowBlock *C = g.newBlock();
  FlowBlock *Q = g.newBlock();
  setIndex(P,0); setIndex(A,1); setIndex(B,2); setIndex(C,3); setIndex(Q,4);

  addLabeledEdge(g,P,A,0x21);
  addLabeledEdge(g,A,B,0x0);
  addLabeledEdge(g,B,C,0x0);
  addLabeledEdge(g,C,Q,0x42);

  Names names;
  names.add(P,"P"); names.add(A,"A"); names.add(B,"B"); names.add(C,"C"); names.add(Q,"Q");
  names.add((const FlowBlock *)&g,"G");	// top-level blocks' parent (addBlock, block.cc:873)

  std::vector<FlowBlock *> nodes;
  nodes.push_back(A);
  nodes.push_back(B);
  FlowBlock *W = g.newBlockList(nodes);
  names.add(W,"W");

  std::cout << "case=cat_basic|top_members=" << names.sortedMembers(g.getList())
            << "|graph_size=" << g.getSize()
            << "|w_index=" << W->getIndex()
            << "|w_children=" << names.sortedMembers(((BlockGraph *)W)->getList())
            << "|parent_A=" << names.parent(A)
            << "|parent_B=" << names.parent(B)
            << "|parent_P=" << names.parent(P)
            << "|flags_A=" << std::hex << A->flags
            << "|flags_B=" << B->flags
            << "|flags_W=" << W->flags
            << "|dead_A=" << ((A->flags & FlowBlock::f_dead) != 0)
            << "|dead_B=" << ((B->flags & FlowBlock::f_dead) != 0)
            << "|w_in=" << edgeList(W,false,names)
            << "|w_out=" << edgeList(W,true,names)
            << "|a_in=" << edgeList(A,false,names)
            << "|a_out=" << edgeList(A,true,names)
            << "|b_in=" << edgeList(B,false,names)
            << "|b_out=" << edgeList(B,true,names)
            << "|p_out=" << edgeList(P,true,names)
            << "|c_in=" << edgeList(C,false,names)
            << "|c_out=" << edgeList(C,true,names)
            << "|q_in=" << edgeList(Q,false,names)
            << "|consistent_W=" << edgesConsistent(W)
            << "|consistent_P=" << edgesConsistent(P)
            << "|consistent_C=" << edgesConsistent(C)
            << "|consistent_A=" << edgesConsistent(A)
            << "|consistent_B=" << edgesConsistent(B)
            << std::dec << std::endl;
}

// Case 2: parallel boundary edges through one peer; selfIdentify's final
// dedup (block.cc:930) must merge the composite's duplicate halves, keep the
// FIRST slot, OR the labels (block.cc:488 eliminateOutDups), and remove the
// peer's mirrored duplicate half.
void caseParallelDedup()
{
  BlockGraph g;
  FlowBlock *P = g.newBlock();
  FlowBlock *A = g.newBlock();
  FlowBlock *B = g.newBlock();
  FlowBlock *M = g.newBlock();
  setIndex(P,0); setIndex(A,1); setIndex(B,2); setIndex(M,3);

  addLabeledEdge(g,P,A,0x1);
  addLabeledEdge(g,A,B,0x0);
  addLabeledEdge(g,A,M,0x5);
  addLabeledEdge(g,B,M,0x9);

  Names names;
  names.add(P,"P"); names.add(A,"A"); names.add(B,"B"); names.add(M,"M");
  names.add((const FlowBlock *)&g,"G");

  std::vector<FlowBlock *> nodes;
  nodes.push_back(A);
  nodes.push_back(B);
  FlowBlock *W = g.newBlockList(nodes);
  names.add(W,"W");

  std::cout << "case=parallel_dedup|w_children=" << names.sortedMembers(((BlockGraph *)W)->getList())
            << "|parent_A=" << names.parent(A)
            << "|parent_B=" << names.parent(B)
            << "|dead_A=" << ((A->flags & FlowBlock::f_dead) != 0)
            << "|dead_B=" << ((B->flags & FlowBlock::f_dead) != 0)
            << "|w_in=" << edgeList(W,false,names)
            << "|w_out=" << edgeList(W,true,names)
            << "|m_in=" << edgeList(M,false,names)
            << "|a_out=" << edgeList(A,true,names)
            << "|b_out=" << edgeList(B,true,names)
            << "|consistent_W=" << edgesConsistent(W)
            << "|consistent_M=" << edgesConsistent(M)
            << std::endl;
}

// Case 3: flag propagation. f_switch_out propagates from a component that
// has an EXTERNAL out edge (selfIdentify block.cc:925-926), but not from a
// fully internal switch-out component; f_interior_gotoout/f_interior_gotoin
// OR-accumulate from every component (identifyInternal block.cc:951).
void caseFlagPropagation()
{
  BlockGraph g;
  FlowBlock *P = g.newBlock();
  FlowBlock *A = g.newBlock();	// switch-out WITH external out edge
  FlowBlock *B = g.newBlock();	// switch-out, external in edge only
  FlowBlock *S = g.newBlock();	// interior goto marks
  FlowBlock *X = g.newBlock();
  FlowBlock *Y = g.newBlock();
  setIndex(P,0); setIndex(A,1); setIndex(B,2); setIndex(S,3); setIndex(X,4); setIndex(Y,5);

  addLabeledEdge(g,P,A,0x0);
  addLabeledEdge(g,A,X,0x0);	// A external out
  addLabeledEdge(g,Y,B,0x0);	// B external in
  addLabeledEdge(g,A,B,0x0);	// internal chain
  addLabeledEdge(g,B,S,0x0);

  A->flags |= FlowBlock::f_switch_out;
  B->flags |= FlowBlock::f_switch_out;	// must NOT propagate (no external out)
  S->flags |= FlowBlock::f_interior_gotoout;
  B->flags |= FlowBlock::f_interior_gotoin;

  Names names;
  names.add(P,"P"); names.add(A,"A"); names.add(B,"B"); names.add(S,"S");
  names.add(X,"X"); names.add(Y,"Y");

  std::vector<FlowBlock *> nodes;
  nodes.push_back(A);
  nodes.push_back(B);
  nodes.push_back(S);
  FlowBlock *W = g.newBlockList(nodes);
  names.add(W,"W");

  std::cout << "case=flag_propagation|flags_W=" << std::hex << W->flags << std::dec
            << "|switch_out_W=" << ((W->flags & FlowBlock::f_switch_out) != 0)
            << "|interior_gotoout_W=" << ((W->flags & FlowBlock::f_interior_gotoout) != 0)
            << "|interior_gotoin_W=" << ((W->flags & FlowBlock::f_interior_gotoin) != 0)
            << "|dead_S=" << ((S->flags & FlowBlock::f_dead) != 0)
            << "|parent_S=" << names.parent(S)
            << "|w_out=" << edgeList(W,true,names)
            << "|w_in=" << edgeList(W,false,names)
            << "|consistent_W=" << edgesConsistent(W)
            << "|consistent_X=" << edgesConsistent(X)
            << "|consistent_Y=" << edgesConsistent(Y)
            << std::endl;
}

// Case 4: self edge inside the consumed set stays a component-internal edge
// and is never inherited by the composite.
void caseSelfEdgeInternal()
{
  BlockGraph g;
  FlowBlock *A = g.newBlock();
  FlowBlock *B = g.newBlock();
  FlowBlock *C = g.newBlock();
  setIndex(A,0); setIndex(B,1); setIndex(C,2);

  addLabeledEdge(g,A,A,0x0);	// self loop within the component set
  addLabeledEdge(g,A,B,0x0);
  addLabeledEdge(g,B,C,0x7);

  Names names;
  names.add(A,"A"); names.add(B,"B"); names.add(C,"C");

  std::vector<FlowBlock *> nodes;
  nodes.push_back(A);
  nodes.push_back(B);
  FlowBlock *W = g.newBlockList(nodes);
  names.add(W,"W");

  std::cout << "case=self_edge_internal|a_in=" << edgeList(A,false,names)
            << "|a_out=" << edgeList(A,true,names)
            << "|w_in=" << edgeList(W,false,names)
            << "|w_out=" << edgeList(W,true,names)
            << "|dead_A=" << ((A->flags & FlowBlock::f_dead) != 0)
            << "|parent_A=" << names.parent(A)
            << "|consistent_A=" << edgesConsistent(A)
            << "|consistent_W=" << edgesConsistent(W)
            << std::endl;
}

} // namespace

int main(void)
{
  std::cout << "schema=1|fixture=BLOCKSTRUCT-IDENTIFY-BOUNDARY-0001"
               "|oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
               "|overall=MISMATCH|covered_projection=MATCH" << std::endl;
  caseCatBasic();
  caseParallelDedup();
  caseFlagPropagation();
  caseSelfEdgeInternal();
  return 0;
}

File: tests/oracle/block_identify_internal_1204.rs
// BLOCKSTRUCT-IDENTIFY-BOUNDARY-0001 Rust comparand for locked Ghidra 12.0.4.
//
// Mirrors block_identify_internal_1204.cc case-for-case through
// CollapseStructure::identify_internal (the BlockGraph::identifyInternal +
// selfIdentify + dedup port, block.cc:940-963 / 895-931 / 525-539) and
// emits the same complete raw state: composite children order, parent
// ownership, raw flags (components are NEVER f_dead), boundary edge
// slots/labels/reverse indices, peer retargets, parallel-edge dedup label
// merge, flag propagation, internal-edge retention.
//
// Top-level positions are normalized to sorted membership: Ghidra appends
// the composite at the list end (block.cc:1695), Rugra installs it at the
// first component's slot — the registered model divergence outside this
// fixture's covered projection.  All other observations are raw.

use std::sync::{Arc, RwLock};

use rugra::block::{block_flags as bf, BlockEdge, BlockGraph, BlockList, BlockType, FlowBlock};
use rugra::blockaction::CollapseStructure;

type BlockRef = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

#[derive(Debug)]
struct BaseProbe {
    index: i32,
    flags: u32,
    incoming: Vec<BlockEdge>,
    outgoing: Vec<BlockEdge>,
}

impl BaseProbe {
    fn new() -> Self {
        Self {
            index: 0,
            flags: 0,
            incoming: Vec::new(),
            outgoing: Vec::new(),
        }
    }
}

impl FlowBlock for BaseProbe {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }

    fn get_index(&self) -> i32 {
        self.index
    }

    fn set_index(&mut self, index: i32) {
        self.index = index;
    }

    fn get_type(&self) -> BlockType {
        BlockType::Plain
    }

    fn get_flags(&self) -> u32 {
        self.flags
    }

    fn set_flags(&mut self, flags: u32) {
        self.flags |= flags;
    }

    fn clear_flags(&mut self, flags: u32) {
        self.flags &= !flags;
    }

    fn size_in(&self) -> usize {
        self.incoming.len()
    }

    fn size_out(&self) -> usize {
        self.outgoing.len()
    }

    fn get_in(&self, slot: usize) -> Option<BlockEdge> {
        self.incoming.get(slot).cloned()
    }

    fn get_out(&self, slot: usize) -> Option<BlockEdge> {
        self.outgoing.get(slot).cloned()
    }

    fn out_edges_mut(&mut self) -> &mut Vec<BlockEdge> {
        &mut self.outgoing
    }

    fn in_edges_mut(&mut self) -> &mut Vec<BlockEdge> {
        &mut self.incoming
    }

    fn add_in_edge(&mut self, edge: BlockEdge) {
        self.incoming.push(edge);
    }

    fn add_out_edge(&mut self, edge: BlockEdge) {
        self.outgoing.push(edge);
    }

    fn get_parent(&self) -> Option<Arc<RwLock<BlockGraph>>> {
        None
    }
}

fn probe(index: i32) -> BlockRef {
    let b: BlockRef = Arc::new(RwLock::new(BaseProbe::new()));
    b.write().unwrap().set_index(index);
    b
}

fn set_label(b: &BlockRef, outgoing: bool, slot: usize, label: u32) {
    let mut g = b.write().unwrap();
    let probe = g.as_any_mut().downcast_mut::<BaseProbe>().unwrap();
    if outgoing {
        probe.outgoing[slot].flags = label;
    } else {
        probe.incoming[slot].flags = label;
    }
}

/// BlockGraph::add_edge mirrors FlowBlock::addInEdge's append ordering and
/// reciprocal slot assignment (self/parallel edges included); mirrored label
/// assignment reproduces the C++ fixture's symmetric addLabeledEdge.
fn wire(graph: &mut BlockGraph, from: &BlockRef, to: &BlockRef, label: u32) {
    graph.add_edge(from.clone(), to.clone());
    let out_slot = from.read().unwrap().size_out() - 1;
    let in_slot = to.read().unwrap().size_in() - 1;
    set_label(from, true, out_slot, label);
    set_label(to, false, in_slot, label);
}

#[derive(Default)]
struct Names {
    blocks: Vec<(BlockRef, String)>,
}

impl Names {
    fn add(&mut self, b: &BlockRef, name: &str) {
        self.blocks.push((b.clone(), name.to_string()));
    }

    fn block(&self, b: Option<&BlockRef>) -> String {
        let Some(b) = b else {
            return "null".to_string();
        };
        self.blocks
            .iter()
            .find(|(c, _)| Arc::ptr_eq(c, b))
            .map_or_else(|| "?".to_string(), |(_, n)| n.clone())
    }

    /// Children of the composite in -nodes- order (Ghidra BlockGraph::list,
    /// block.hh:359; Rugra's BlockList.children holds the same Arc nodes).
    fn children(&self, composite: &BlockRef) -> Vec<BlockRef> {
        let g = composite.read().unwrap();
        g.as_any()
            .downcast_ref::<BlockList>()
            .map(|l| l.children.clone())
            .unwrap_or_default()
    }

    /// Parent ownership mirror of Ghidra FlowBlock::getParent (block.hh:78):
    /// a component of the composite resolves to the composite, a top-level
    /// block to the graph itself.
    fn parent(&self, b: &BlockRef, composite: &BlockRef, graph_blocks: &[BlockRef]) -> String {
        for c in &self.children(composite) {
            if Arc::ptr_eq(c, b) {
                return self.block(Some(composite));
            }
        }
        for g in graph_blocks {
            if Arc::ptr_eq(g, b) {
                return "G".to_string();
            }
        }
        "?".to_string()
    }

    fn sorted_members(&self, blocks: &[BlockRef]) -> String {
        let mut names: Vec<String> = blocks.iter().map(|b| self.block(Some(b))).collect();
        names.sort();
        format!("[{}]", names.join(","))
    }
}

fn edge_list(b: &BlockRef, outgoing: bool, names: &Names) -> String {
    let g = b.read().unwrap();
    let n = if outgoing { g.size_out() } else { g.size_in() };
    let mut parts: Vec<String> = Vec::new();
    for slot in 0..n {
        let e = if outgoing { g.get_out(slot) } else { g.get_in(slot) };
        if let Some(e) = e {
            // Decimal label formatting mirrors the C++ fixture's fresh
            // ostringstream inside edgeList (its own decimal stream state).
            parts.push(format!(
                "{}:{}:{}",
                names.block(Some(&e.point)),
                e.reverse_index,
                e.flags
            ));
        }
    }
    format!("[{}]", parts.join(","))
}

/// Reciprocal-consistency mirror of FlowBlock::checkEdges (block.cc:545-570).
fn edges_consistent(b: &BlockRef) -> bool {
    let g = b.read().unwrap();
    for i in 0..g.size_in() {
        let Some(e) = g.get_in(i) else { continue };
        let peer = e.point.read().unwrap();
        let rev = e.reverse_index;
        if rev < 0 || peer.size_out() as i32 <= rev {
            return false;
        }
        let Some(pe) = peer.get_out(rev as usize) else { return false };
        if !Arc::ptr_eq(&pe.point, b) || pe.reverse_index != i as i32 {
            return false;
        }
    }
    for i in 0..g.size_out() {
        let Some(e) = g.get_out(i) else { continue };
        let peer = e.point.read().unwrap();
        let rev = e.reverse_index;
        if rev < 0 || peer.size_in() as i32 <= rev {
            return false;
        }
        let Some(pe) = peer.get_in(rev as usize) else { return false };
        if !Arc::ptr_eq(&pe.point, b) || pe.reverse_index != i as i32 {
            return false;
        }
    }
    true
}

/// Top-level members (non-consumed) in slot order plus their count — the
/// observable Ghidra's incremental list compaction (block.cc:953-960)
/// maintains; membership comes from the absorbed_into parent record.
fn top_members(graph: &BlockGraph) -> Vec<BlockRef> {
    let consumed = &graph.absorbed_into;
    let mut out = Vec::new();
    for i in 0..graph.get_size() {
        if let Some(b) = graph.get_block(i) {
            let idx = b.read().unwrap().get_index();
            if !consumed.contains_key(&idx) {
                out.push(b);
            }
        }
    }
    out
}

fn flags_of(b: &BlockRef) -> String {
    format!("{:x}", b.read().unwrap().get_flags())
}

fn dead(b: &BlockRef) -> u32 {
    (b.read().unwrap().get_flags() & bf::DEAD != 0) as u32
}

// Case 1: plain cat identify (newBlockList, block.cc:1758-1774).
fn case_cat_basic() {
    let p = probe(0);
    let a = probe(1);
    let b = probe(2);
    let c = probe(3);
    let q = probe(4);
    let mut graph = BlockGraph::new();
    for blk in [&p, &a, &b, &c, &q] {
        graph.blocks.push((*blk).clone());
    }
    wire(&mut graph, &p, &a, 0x21);
    wire(&mut graph, &a, &b, 0x0);
    wire(&mut graph, &b, &c, 0x0);
    wire(&mut graph, &c, &q, 0x42);

    // newBlockList({A,B}): A is the install block (Ghidra nodes[0]), B the
    // consumed component — the try_rule_cat split of Ghidra's -nodes-.
    let w: BlockRef = Arc::new(RwLock::new(BlockList::new(1, vec![a.clone(), b.clone()])));
    {
        let mut collapse = CollapseStructure::new(&mut graph, "identify_fixture");
        collapse.identify_internal(&w, &[2], 1);
    }

    let mut names = Names::default();
    names.add(&p, "P");
    names.add(&a, "A");
    names.add(&b, "B");
    names.add(&c, "C");
    names.add(&q, "Q");
    names.add(&w, "W");

    let members = top_members(&graph);
    println!(
        "case=cat_basic|top_members={}|graph_size={}|w_index={}|w_children={}|parent_A={}|parent_B={}|parent_P={}|flags_A={}|flags_B={}|flags_W={}|dead_A={}|dead_B={}|w_in={}|w_out={}|a_in={}|a_out={}|b_in={}|b_out={}|p_out={}|c_in={}|c_out={}|q_in={}|consistent_W={}|consistent_P={}|consistent_C={}|consistent_A={}|consistent_B={}",
        names.sorted_members(&members),
        members.len(),
        w.read().unwrap().get_index(),
        names.sorted_members(&names.children(&w)),
        names.parent(&a, &w, &graph.blocks),
        names.parent(&b, &w, &graph.blocks),
        names.parent(&p, &w, &graph.blocks),
        flags_of(&a),
        flags_of(&b),
        flags_of(&w),
        dead(&a),
        dead(&b),
        edge_list(&w, false, &names),
        edge_list(&w, true, &names),
        edge_list(&a, false, &names),
        edge_list(&a, true, &names),
        edge_list(&b, false, &names),
        edge_list(&b, true, &names),
        edge_list(&p, true, &names),
        edge_list(&c, false, &names),
        edge_list(&c, true, &names),
        edge_list(&q, false, &names),
        edges_consistent(&w) as u32,
        edges_consistent(&p) as u32,
        edges_consistent(&c) as u32,
        edges_consistent(&a) as u32,
        edges_consistent(&b) as u32,
    );
}

// Case 2: parallel boundary edges through one peer; selfIdentify's dedup
// (block.cc:930) merges duplicate halves, keeps the first slot, ORs labels.
fn case_parallel_dedup() {
    let p = probe(0);
    let a = probe(1);
    let b = probe(2);
    let m = probe(3);
    let mut graph = BlockGraph::new();
    for blk in [&p, &a, &b, &m] {
        graph.blocks.push((*blk).clone());
    }
    wire(&mut graph, &p, &a, 0x1);
    wire(&mut graph, &a, &b, 0x0);
    wire(&mut graph, &a, &m, 0x5);
    wire(&mut graph, &b, &m, 0x9);

    let w: BlockRef = Arc::new(RwLock::new(BlockList::new(1, vec![a.clone(), b.clone()])));
    {
        let mut collapse = CollapseStructure::new(&mut graph, "identify_fixture");
        collapse.identify_internal(&w, &[2], 1);
    }

    let mut names = Names::default();
    names.add(&p, "P");
    names.add(&a, "A");
    names.add(&b, "B");
    names.add(&m, "M");
    names.add(&w, "W");

    println!(
        "case=parallel_dedup|w_children={}|parent_A={}|parent_B={}|dead_A={}|dead_B={}|w_in={}|w_out={}|m_in={}|a_out={}|b_out={}|consistent_W={}|consistent_M={}",
        names.sorted_members(&names.children(&w)),
        names.parent(&a, &w, &graph.blocks),
        names.parent(&b, &w, &graph.blocks),
        dead(&a),
        dead(&b),
        edge_list(&w, false, &names),
        edge_list(&w, true, &names),
        edge_list(&m, false, &names),
        edge_list(&a, true, &names),
        edge_list(&b, true, &names),
        edges_consistent(&w) as u32,
        edges_consistent(&m) as u32,
    );
}

// Case 3: flag propagation (f_switch_out needs an external out edge,
// block.cc:925-926; f_interior_goto* OR from every component, block.cc:951).
fn case_flag_propagation() {
    let p = probe(0);
    let a = probe(1);
    let b = probe(2);
    let s = probe(3);
    let x = probe(4);
    let y = probe(5);
    let mut graph = BlockGraph::new();
    for blk in [&p, &a, &b, &s, &x, &y] {
        graph.blocks.push((*blk).clone());
    }
    wire(&mut graph, &p, &a, 0x0);
    wire(&mut graph, &a, &x, 0x0);
    wire(&mut graph, &y, &b, 0x0);
    wire(&mut graph, &a, &b, 0x0);
    wire(&mut graph, &b, &s, 0x0);

    a.write().unwrap().set_flags(bf::SWITCH_OUT);
    b.write().unwrap().set_flags(bf::SWITCH_OUT | bf::INTERIOR_GOTOIN);
    s.write().unwrap().set_flags(bf::INTERIOR_GOTOOUT);

    let w: BlockRef = Arc::new(RwLock::new(BlockList::new(
        1,
        vec![a.clone(), b.clone(), s.clone()],
    )));
    {
        let mut collapse = CollapseStructure::new(&mut graph, "identify_fixture");
        collapse.identify_internal(&w, &[2, 3], 1);
    }

    let mut names = Names::default();
    names.add(&p, "P");
    names.add(&a, "A");
    names.add(&b, "B");
    names.add(&s, "S");
    names.add(&x, "X");
    names.add(&y, "Y");
    names.add(&w, "W");

    let wf = w.read().unwrap().get_flags();
    println!(
        "case=flag_propagation|flags_W={:x}|switch_out_W={}|interior_gotoout_W={}|interior_gotoin_W={}|dead_S={}|parent_S={}|w_out={}|w_in={}|consistent_W={}|consistent_X={}|consistent_Y={}",
        wf,
        (wf & bf::SWITCH_OUT != 0) as u32,
        (wf & bf::INTERIOR_GOTOOUT != 0) as u32,
        (wf & bf::INTERIOR_GOTOIN != 0) as u32,
        dead(&s),
        names.parent(&s, &w, &graph.blocks),
        edge_list(&w, true, &names),
        edge_list(&w, false, &names),
        edges_consistent(&w) as u32,
        edges_consistent(&x) as u32,
        edges_consistent(&y) as u32,
    );
}

// Case 4: self edge inside the consumed set stays component-internal.
fn case_self_edge_internal() {
    let a = probe(0);
    let b = probe(1);
    let c = probe(2);
    let mut graph = BlockGraph::new();
    for blk in [&a, &b, &c] {
        graph.blocks.push((*blk).clone());
    }
    wire(&mut graph, &a, &a, 0x0);
    wire(&mut graph, &a, &b, 0x0);
    wire(&mut graph, &b, &c, 0x7);

    let w: BlockRef = Arc::new(RwLock::new(BlockList::new(0, vec![a.clone(), b.clone()])));
    {
        let mut collapse = CollapseStructure::new(&mut graph, "identify_fixture");
        collapse.identify_internal(&w, &[1], 0);
    }

    let mut names = Names::default();
    names.add(&a, "A");
    names.add(&b, "B");
    names.add(&c, "C");
    names.add(&w, "W");

    println!(
        "case=self_edge_internal|a_in={}|a_out={}|w_in={}|w_out={}|dead_A={}|parent_A={}|consistent_A={}|consistent_W={}",
        edge_list(&a, false, &names),
        edge_list(&a, true, &names),
        edge_list(&w, false, &names),
        edge_list(&w, true, &names),
        dead(&a),
        names.parent(&a, &w, &graph.blocks),
        edges_consistent(&a) as u32,
        edges_consistent(&w) as u32,
    );
}

fn main() {
    println!(
        "schema=1|fixture=BLOCKSTRUCT-IDENTIFY-BOUNDARY-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b|overall=MISMATCH|covered_projection=MATCH"
    );
    case_cat_basic();
    case_parallel_dedup();
    case_flag_propagation();
    case_self_edge_internal();
}
